### PR TITLE
[RFC]Rockchip: add SRAM section in kern.ld.S

### PR DESCRIPTION
**Pull request by TonyXie06**
I have some code and data must be put on SRAM for some requirements,  I refer to implementation on arm-trusted-firmware[0] and try to add a SRAM section in kern.ld.S. But: 
1. I am not sure whether it's a good way or not, if not, I wish you experts can give me some advice.
2. for these patches, I meet a problem while compiling:  I add a foo() function on SRAM for test, but there comes error compile log: **out/arm-plat-rockchip/core/arch/arm/plat-rockchip/psci_rk322x.o:(.ARM.exidx.sram.text+0x0): relocation truncated to fit: R_ARM_PREL31 against `.sram.text'**  I tried, I have to ask for your experts help.

[0]: web link:
https://github.com/ARM-software/arm-trusted-firmware/blob/master/plat/rockchip/rk3328/include/plat.ld.S 
https://github.com/ARM-software/arm-trusted-firmware/blob/master/bl31/bl31.ld.S

Signed-off-by: Joseph Chen <chenjh@rock-chips.com>

### CODE REVIEW

File: core/arch/arm/plat-rockchip/common.h [hunk 1]
```diff
@@ -46,4 +46,7 @@
 #define BITS_WITH_WMASK(bits, msk, shift) \
 				(SHIFT_U32(bits, shift) | BITS_WMSK(msk, shift))
 
+#define __sramfunc __attribute__((section(".sram.text"))) \
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-rockchip/common.h]: This is not 100% proof. Static data allocated by the function may still end up in `.rodata`, `.data` or `.bss`.

The technique used to separate unpaged and init code/rodata from paged code/rodata can be used here too.

File: core/arch/arm/plat-rockchip/psci_rk322x.c [hunk 2]
```diff
@@ -377,12 +377,18 @@ int psci_system_suspend(uintptr_t entry __unused,
 	return PSCI_RET_SUCCESS;
 }
 
+__sramfunc static void foo(void)
+{
+	IMSG("This is a foo function test for sram section");
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-rockchip/psci_rk322x.c]: The string argument for `IMSG()` is likely not stored in `.sram`.

File: core/arch/arm/kernel/kern.ld.S [hunk 3]
```diff
@@ -394,6 +394,12 @@ SECTIONS
 	__init_size = __data_end - CFG_TEE_LOAD_ADDR;
 	__init_mem_usage = __end - CFG_TEE_LOAD_ADDR;
 #endif
+
+#ifdef CFG_PLAT_EXTRA_LD_SCRIPT
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/kern.ld.S]: I'd rather have a good generic solution to the problem you're trying to solve.

File: core/arch/arm/plat-rockchip/common.h (hunk 1 above)
Comment (TonyXie06) [inline core/arch/arm/plat-rockchip/common.h]: Sorry, I miss the definition here. I will use them for sram data, rodata.
#define __sramdata __attribute__((section(".sram.data")))
#define __sramconst __attribute__((section(".sram.rodata")))

File: core/arch/arm/plat-rockchip/psci_rk322x.c (hunk 2 above)
Comment (TonyXie06) [inline core/arch/arm/plat-rockchip/psci_rk322x.c]: my mistake.

File: core/arch/arm/kernel/kern.ld.S (hunk 3 above)
Comment (TonyXie06) [inline core/arch/arm/kernel/kern.ld.S]: How about include plat.ld.S in core/arch/arm/plat-rockchip/kern.ld.S, like this: 
#include "../kernel/kern.ld.S"
#include "plat.ld.S"

**Comment by TonyXie06** [comment]
You adviced me compile with -mlong-calls, but it seems doesn't work, the complie log is still error. output is same.
I added in core/arch/arm/plat-rockchip/sub.mk, like this: 

 ifeq ($(PLATFORM_FLAVOR),rk322x)
 srcs-$(CFG_PSCI_ARM32) += psci_rk322x.c
**cflags-psci_rk322x.c-(CFG_PSCI_ARM32) += -mlong-calls**
 endif

I tried and find 2 ways that  compile can success:
1. increase ISRAM_LDS_BASE. It seems this match your thought "the link error is probably because the jump to sram is larger than what can be jumped with a bx <immediate>."
2. remove -funwind-tables. I know it is not suitable way, it's just a test.

File: core/arch/arm/kernel/kern.ld.S (hunk 3 above)
Comment (TonyXie06) [inline core/arch/arm/kernel/kern.ld.S]: This can avoid touching  generic kern.ld.S, only platform kern.ld.S be touched.

**Comment by jenswi-linaro** [comment]
Try appling `-mlong-calls` on all c files instead, perhaps that helps.

I don't understand what you're trying to achieve here, "some requirements" is a bit vague. Using only these `__sram*` annotation will only be 100% for trivial functions. When you start more complicated functions you'll run into surprises now and then. It's simply not robust if you need 100% coverage.

**Comment by TonyXie06** [comment]
"some requirements" are these on rockchip platforms:
1.  when system suspend, some platforms support ddr self refresh by hardware while some not. if not support , the cpu0 should run on  SRAM to make ddr self refresh and execute other remaining steps that mainly for low power.
2.  for ddr frequency scale,  we usually let nonboot cpus run on SRAM and stop in wfi/e state,  then boot cpu will execute the ddr frequency scale code on SRAM;
3.  for some platfoms, last cpu will lose power when system suspend, the resume code should be put on SRAM.

In linux kernel, it provides PIE for us to solve these requirements, but as we introduce optee on our platforms, we move the solution of these requirements into optee. If there is not a SRAM section, maybe it's hard for us to implement relevant code for requirements , we maybe need write assembly code and make it position independent executables. So you can see that we add a plat.ld.S in arm-trusted-firmware too.

**Comment by jenswi-linaro** [comment]
OK, I'm fine with the changes you propose in `core/arch/arm/plat-rockchip/kern.ld.S` etc.
I assume that you will test that the concept works before we start merging stuff related to this.

**Comment by TonyXie06** [comment]
Hi jens, I applied -mlong-calls in /core/core.mk like below, but the compile log was still error.

diff --git a/core/core.mk b/core/core.mk
index bcf788d..468dcf7 100644
--- a/core/core.mk
+++ b/core/core.mk
@@ -26,7 +26,7 @@ cppflags$(sm) += -Icore/include
 cppflags$(sm)  += -include $(conf-file)
 cppflags$(sm)  += -I$(out-dir)/core/include/generated
 cppflags$(sm)  += $(core-platform-cppflags)
-cflags$(sm)    += $(core-platform-cflags)
**+cflags$(sm)    += $(core-platform-cflags) -mlong-calls**
 ifeq ($(CFG_CORE_SANITIZE_UNDEFINED),y)
 cflags$(sm)    += -fsanitize=undefined
 endif

**Comment by jenswi-linaro** [comment]
Compiling with `CFG_UNWIND=n` seems to help.

**Comment by TonyXie06** [comment]
Thanks, compile successfully !

**Comment by mmind** [comment]
More like a general comment on this. 

The arm-trusted-firmware used on arm64 Rockchip socs as base, seems to get armv7 support now [https://github.com/ARM-software/arm-trusted-firmware/pull/1145], which in turn can carry OP-TEE as payload.

So to unify the 32bit and 64bit approach, maybe it might be nicer to actually move the hardware-specific parts over there, so that we don't have multiple instances of things like PSCI?

**Comment by jenswi-linaro** [comment]
In the ARMv7 case ARM-TF is only a loader, it's completely gone once the boot is done unless I'm missing something.

**Comment by mmind** [comment]
Hmm, looking at a platform-branch which seems to be meant to go on top of that core armv7 work [https://github.com/etienne-lms/arm-trusted-firmware/commit/81ed61cb3dd32c7964440040e4491e5da02db7e3#diff-ad6924488efce0d9cd3751d7b575a5d8R2] it looks like the developer is planning on doing PSCI at the ATF level.

I don't claim on knowing all the details, I just want to make sure we don't needlessly duplicate ATF functionality.

**Comment by etienne-lms** [comment]
> In the ARMv7 case ARM-TF is only a loader, it's completely gone once the boot is done unless I'm missing something.

Agree. When loading an secure firmware (I.e op-tee) on arm-tf/armv7, only the boot stages of arm-tf are used, no runtime services. All runtime services (i.e psci) shall be provided by the secure firmware (op-tee).

**Comment by etienne-lms** [comment]
@mmind: psci support in the arm-tf would be for cases where there is no op-tee loaded. In this case, arm-tf would provide the required run-time serivce, i.e PSCI.

**Comment by mmind** [comment]
Ok, thanks for the clarification ... then just ignore my comments :-)

**Comment by mmind** [comment]
@etienne-lms Ok, thanks for the clarification ... then just ignore my comments :-)

Just for my personal education, is there a specific reason for this, in contrast to keeping ATF parts around on 64bit?

I.e. on 64bit Rockchip platforms, the ATF does things like PSCI but also DDR frequency scaling and while the IP blocks involved are not identical they are still similar in a lot of cases. So having that functionality in both feels somewhat counter-intuitive on first glance.

**Comment by etienne-lms** [comment]
You're right but armv7 does not offer nice hw to support 2 runtime firmwares in the trustzone secure world, while aarch64 capable armv8 do through EL3/EL1-S.
One who just wants an armv7 secure bootloader for loading the optee could rely on arm-tf boot stage and don't care about psci support in arm-tf. And of course implements all runtime secure services within the op-tee (i.e psci).

**Pull request closed without merging**